Consolidation of tests in regression test suite

paulromano opened the pull request:
This pull requests consolidates many tests (partially addressing #262), in total reducing the number of tests from 102 to 64. This should help shorten the runs on Travis CI and allow us a little more room to grow the test suite. The main changes I made were:
- Consolidated most of the `test_filter_*` and `test_score_*` tests into a single `test_tallies` test. All these tests were using the same geometry, materials, and settings, so combining them into a single test with lots of tallies results in the same coverage but much less run time.
- Combined four `test_density_*` tests into a single one that covers all cases.
- Combined four `test_plot_*` tests into a single one. I've also expanded it a little bit because I realized we have no test that produces a voxel plot.
- Combined two `test_salphabeta*` tests into a single one
- Removed test_basic -- doesn't add anything in terms of coverage.

I also made a few changes that allow code coverage to easily be generated (that doesn't rely explicitly on ctest/cdash. So long as you have gcov/lcov installed, you can run `./run_tests.py -C ^hdf5-debug_coverage -s` and at the end you'll see a new directory called "coverage" which includes HTML pages showing details on code coverage. By my measure, our code coverage has increased from 69.2% to 70.2% due to this PR.

I still need to confirm that:
- [x] I haven't completely broken valgrind memory testing

smharper commented on tests/test_plot/test_plot.py:
```diff
@@ -0,0 +1,71 @@
+                outstr += fh['num_voxels'].value.tobytes()
+                outstr += fh['lower_left'].value.tobytes()
+                outstr += fh['voxel_width'].value.tobytes()
+                outstr += fh['data'].value.tobytes()
```
I get a fail with `AttributeError: 'numpy.ndarray' object has no attribute 'tobytes'`.  Version issue?  Just change these `.tostring` and that should be fixed

smharper commented:
I really like this PR!  The local coverage is a feature I've really been wanting.

Short comments:
* Roll up "test_many_scores" into "test_tallies"
* Add a tally with `tally.nuclides = 'all'` to "test_tallies"
* Add a plot mask to "test_plots"
* We should probably make another "test_tallies" for survival biasing
* Modify the "test_tallies" results to include a print-out of all the tallies in the summary file using the Python API

Sticking all of our scores and filters into one test does go against standard testing wisdom that every test should be atomic and you should know what's failing by the name alone.  But I think it's okay in this case because you can tell what's failing by the tally ID in results_error (and because in my experience, you usually break most of the tallies if you break any).

smharper commented:
Add *.voxel and tests/coverage to .gitignore

paulromano commented:
@smharper I've addressed all your comments, except for the one about including a printout of tallies from summary.h5. I'm not 100% sure what you want, but I will say that if I try to include `repr(summary.tallies)`, it's going to be different between Python 2 and 3. I should also mention that test_plot already included a mask, so I've haven't added anything extra.

I've also improved/fixed our memory leak checking ability which uses valgrind. You can now run `./run_tests.py -C ^hdf5-debug_valgrind -s` and at the end you'll get a directory "memcheck" with the results of running valgrind on a select set of tests. The way it works is a little tricky -- we can't run `valgrind ... python test_blahblah.py` because then valgrind is effectively checking python, not OpenMC. Now, when CMakeLists.txt sees that it's a memcheck run, it uses the fancy --build-inputs option to build input files, and then instructs the tests to run OpenMC directly. It was set up to do this before, but as the test suite has evolved over time the valgrind capability hasn't really kept up. I should note that while valgrind does currently report memory leaks, when I run it accounting for the changes in #557, there are no leaks!

smharper commented:
Ah too bad about the summary.  I was hoping that we could make it work by using sorted keys on the `summary.tallies` dictionary, but I see now that the Python 2 strings have a `u` tag and the Python3 ones do not.

Also, wow test_tallies is an expensive test.  It's good to keep that test in mind when you hear people saying that cross section lookup is the most expensive part of MC.  I want to run some profiling on that test sometime.

Good PR.  Merging

paulromano commented:
Thanks for reviewing and helping to improve with your suggestions. Yes, tally performance is a big deal in my opinion and often takes a back seat to cross sections. Something to keep in mind as we attack #214.
